### PR TITLE
Modify query parsing to require all tokens

### PR DESCRIPTION
**Pull request by joelverhagen**
## Related issues

https://github.com/NuGet/NuGetGallery/issues/8130 - Improve search: Re-evaluate camel-split strategy
https://github.com/NuGet/NuGetGallery/issues/8201 - Poor/unrelated search results when not using the exact package name
_(via email)_ - Microsoft.UI.Xaml (a.k.a. WinUI) package is hard to find
_(via email)_ - Microsoft.Windows.SDK.BuildTools package is hard to find

## Reasoning

There have been several reports that made me question our approach to camel-split. After much experimentation and prototyping, I found that the main problem is that camel-split produces so many generic tokens that the results set for the user query is often massive and therefore full on irrelevant packages.

My theory is that our search will be more useful if we require that **ALL** provided tokens are in the matched package's metadata. Currently, we use the default Lucene OR-like behavior. The behavior is that documents matching all tokens are boosted but documents only matching a single token are still included. This Lucene behavior is fine but, in conjunction with our massive download boost, it results in hardly relevant packages yet popular packages at the top.

This change to require all tokens comes at a cost though. Currently, the user can get "sibling" results pretty easily even when they have a very specific search. If, say, 8 of the 9 tokens are matched the package will still be pretty close to the top. However my feeling is that if a user wants more generic results, they can easily modify their search query to be more generic.

In other words, we should at least give the user the option to make a specific search query. If the result set is too small, they can refine their search query to be less specific. Currently, this incremental refinement is very difficult or impossible.

## Steps

- [ ] Merge this PR
- [ ] Deploy this change to the preview search endpoints. **No index rebuild is required.**
- [ ] Start an A/B test on nuget.org
- [ ] Verify the NuGet.org search KPIs improved
- [ ] Ship the change to the primary search endpoints
- [ ] Verify the Visual Studio search KPIs improved
- [ ] Merge the `ab-cameljoin` branch into `dev` branch

## Search scorer output

Variant                        | Control | Treatment | Delta
------------------------------ | ------- | --------- | -------
NuGet.org curated searches     | 0.7639  | 0.7583    | ⚠ -0.0054
Visual Studio curated searches | 0.6827  | 0.6923    | ✔ +0.0096
Feedback curated search        | 0.6760  | 0.7067    | ✔ +0.0240

Note that the lion's share of the NuGet.org curated search score decrease is due to the exact match sibling case. As described above, I think this is the right move.

<details>
<summary>Full tool output</summary>

    Curated Search Queries
    ======================
    Control:   0.763790325811432
    Treatment: 0.758343892040208

    Biggest Winners (10)
    --------------------
    excel                          => +0.0025
    pdf                            => +0.0023
    ef                             => +0.0019
    devexpress                     => +0.0013
    entity                         => +0.0011
    csv                            => +0.0011
    swagger                        => +0.0009
    Microsoft.AspNet.WebApi.Client => +0.0006
    System.Web.Http                => +0.0005
    entity framework               => +0.0005

    Biggest Losers (10)
    -------------------
    Microsoft.AspNetCore.Authentication.JwtBearer       => -0.0020
    Microsoft.AspNetCore.Mvc.NewtonsoftJson             => -0.0019
    sqlite                                              => -0.0012
    Microsoft.NET.Test.Sdk                              => -0.0011
    Microsoft.EntityFrameworkCore.Design                => -0.0010
    System.Threading.Tasks.Extensions                   => -0.0009
    AutoMapper.Extensions.Microsoft.DependencyInjection => -0.0009
    moq                                                 => -0.0008
    sql                                                 => -0.0007
    xunit.runner.visualstudio                           => -0.0006

    Lowest Treatment Scores (10)
    ----------------------------
    blazor                    => 0
    system.web                => 0
    system.web.mvc            => 0
    System.Web                => 0
    System.Web.Mvc            => 0
    Microsoft.Practices.Unity => 0
    System.Web.Optimization   => 0
    Tags:"Blazor"             => 0
    Bluetooth                 => 0
    ColorPicker               => 0

    Client Curated Search Queries
    =============================
    Control:   0.682689456302459
    Treatment: 0.692252985191534

    Biggest Winners (10)
    --------------------
    entity  => +0.0046
    en      => +0.0029
    e       => +0.0022
    Entity  => +0.0015
    c       => +0.0012
    ef      => +0.0012
    En      => +0.0012
    swagger => +0.0011
    E       => +0.0010
    entit   => +0.0010

    Biggest Losers (10)
    -------------------
    Microsoft.AspNetCore.Mvc.NewtonsoftJson       => -0.0028
    sql                                           => -0.0021
    Microsoft.AspNetCore.Authentication.JwtBearer => -0.0013
    s                                             => -0.0012
    Microsoft.EntityFrameworkCore.Design          => -0.0011
    moq                                           => -0.0011
    Microsoft.EntityFrameworkCore.Tools           => -0.0010
    sqlite                                        => -0.0009
    microsoft.aspnet                              => -0.0009
    Microsoft.Extensions.Configuration.Json       => -0.0008

    Lowest Treatment Scores (10)
    ----------------------------
    system.web      => 0
    System.Web      => 0
    boot            => 0
    system.web.http => +0.0001
    aspnetcore.app  => +0.0002
    system.buffers  => +0.0002
    test            => +0.0002
    SignalR         => +0.0003
    Micros          => +0.0003
    SQLite          => +0.0003

    Feedback
    ========
    Control:   0.675991835041447
    Treatment: 0.706677499081473

    Biggest Winners (10)
    --------------------
    Azure                                       => +0.0024
    extended xml                                => +0.0024
    microsoft hosting windowsservice            => +0.0024
    microsoft.extensions.hosting.windowsservice => +0.0024
    Microsoft.Extensions.Hosting.WindowsService => +0.0024
    my sample lib                               => +0.0024
    System.IO.Abstractions.Test                 => +0.0024
    uwp pdf                                     => +0.0024
    ServerHost                                  => +0.0020
    extensions logging log4net                  => +0.0018

    Biggest Losers (10)
    -------------------
    aspnetcore app                     => -0.0024
    MicrosoftEntityFrameWorkCore.Tools => -0.0024
    CK                                 => -0.0018
    common mark                        => -0.0018
    micro orm                          => -0.0018
    aspnet                             => -0.0014
    cosmos                             => -0.0014
    EntityFramework                    => -0.0014
    FluentAssertion                    => -0.0014
    MADE                               => -0.0014

    Lowest Treatment Scores (10)
    ----------------------------
    azure     => 0
    log file  => 0
    log files => 0
    log       => 0
    logging   => 0
    azure f#  => 0
    log       => 0
    logging   => 0
    Enity     => 0
    Wpf       => 0
</details>

## Try it out!

You can type queries into the PROD-PERF environment, like this:
https://azuresearch-usnc-perf.nuget.org/query?q=WinUI

## Other notes

I also fixed the description analyzer to lowercase tokens before the stopword filtering. This did not move the score at all, so I will not rebuild the index just for this purpose. Next time we have an index rebuild, this change will automatically be included.

### Refinement flow

For example, consider the search for `Microsoft.AspNetCore.Authentication.JwtBearer`.

My proposal yields this:
<pre>
Rank | <a href="https://www.nuget.org/packages?q=Microsoft.AspNetCore.Authentication.JwtBearer">Control (73.9k hits)</a>                                       | Treatment (1 hits)
---- | ---------------------------------------------------------- | ----------------------------------------------------------
1    | Microsoft.AspNetCore.Authentication.JwtBearer (71.66M)     | Microsoft.AspNetCore.Authentication.JwtBearer (71.08M)
2    | Microsoft.AspNetCore.Authentication.Core (141.81M)         |
3    | Microsoft.AspNetCore.Authentication.Abstractions (144.86M) |
4    | Microsoft.AspNetCore.Routing (174.81M)                     |
5    | Microsoft.AspNetCore.Authentication (71.77M)               |
6    | Microsoft.AspNetCore.Http (279.99M)                        |
7    | Microsoft.AspNetCore.Http.Features (336.17M)               |
8    | Microsoft.AspNetCore.Http.Abstractions (323.48M)           |
9    | Microsoft.AspNetCore.Hosting.Server.Abstractions (310.39M) |
10   | Microsoft.AspNetCore.Hosting.Abstractions (310.22M)        |
</pre>

If the user wants the other authentication packages, they can easily modify their search result to `Microsoft.AspNetCore.Authentication`.

<pre>
Rank | <a href="https://www.nuget.org/packages?q=Microsoft.AspNetCore.Authentication">Control (73.73k hits)</a>                                         | Treatment (70 hits)
---- | ------------------------------------------------------------- | -------------------------------------------------------------
1    | Microsoft.AspNetCore.Authentication (71.77M)                  | Microsoft.AspNetCore.Authentication (71.43M)
2    | Microsoft.AspNetCore.Authentication.Core (141.81M)            | Microsoft.AspNetCore.Authentication.Abstractions (143.93M)
3    | Microsoft.AspNetCore.Authentication.Abstractions (144.86M)    | Microsoft.AspNetCore.Authentication.Core (140.9M)
4    | Microsoft.AspNetCore.Authentication.JwtBearer (71.66M)        | Microsoft.AspNetCore.Authentication.JwtBearer (71.08M)
5    | Microsoft.AspNetCore.Authentication.Cookies (47.81M)          | Microsoft.AspNetCore.Authentication.OpenIdConnect (49.88M)
6    | Microsoft.AspNetCore.Authentication.OpenIdConnect (50.17M)    | Microsoft.AspNetCore.Authentication.Cookies (47.62M)
7    | Microsoft.AspNetCore.Authentication.MicrosoftAccount (32.15M) | Microsoft.AspNetCore.Authentication.OAuth (40.07M)
8    | Microsoft.AspNetCore.Authentication.OAuth (40.2M)             | Microsoft.AspNetCore.Authentication.MicrosoftAccount (32.04M)
9    | Microsoft.AspNetCore.Authentication.Google (37.68M)           | Microsoft.AspNetCore.Authentication.Google (37.56M)
10   | Microsoft.AspNetCore.Authentication.Facebook (32.56M)         | Microsoft.AspNetCore.Authentication.Facebook (32.45M)
</pre>

#### `WinUI`

The `WinUI` search is very likely for the popular `Microsoft.UI.Xaml` or perhaps the `Uno.WinUI.*` packages.

<pre>
Rank | <a href="https://www.nuget.org/packages?q=WinUI">Control (5.8k hits)</a>                  | Treatment (233 hits)
---- | ------------------------------------ | ------------------------------------
1    | Microsoft.Win32.Registry (279.16M)   | Microsoft.UI.Xaml (920.89k)
2    | Microsoft.Win32.Primitives (160.91M) | Uno.WinUI.Lottie (27.57k)
3    | Uno.WinUI.Lottie (28.56k)            | Uno.WinUI (30.16k)
4    | Uno.WinUI (31.11k)                   | Uno.WinUI.Maps (27.67k)
5    | Uno.WinUI.Maps (28.7k)               | Uno.WinUI.MSAL (24.11k)
6    | Uno.WinUI.MSAL (25.01k)              | Uno.WinUI.DualScreen (27.86k)
7    | Uno.WinUI.RemoteControl (30.92k)     | Uno.WinUI.RemoteControl (29.94k)
8    | Avalonia.Win32 (225.7k)              | Uno.WinUI.WebAssembly (19.64k)
9    | Uno.WinUI.DualScreen (28.83k)        | Uno.WinUI.Runtime.WebAssembly (21.23k)
10   | HtmlRenderer.WinForms (234.17k)      | Uno.WinUI.Skia.Gtk (12.66k)
</pre>

#### `Microsoft Windows SDK BuildTools`

The `Microsoft Windows SDK BuildTools` is very specific, in the sense that many search terms provides a lot of specificity, but this returns a whopping 30 thousand packages in PROD today. With download boosting approach, this yields useless results.

<pre>
Rank | <a href="https://www.nuget.org/packages?q=Microsoft+Windows+SDK+BuildTools">Control (30.13k hits)</a>                                                 | Treatment (6 hits)
---- | --------------------------------------------------------------------- | ---------------------------------------------------------------------
1    | Microsoft.Net.UWPCoreRuntimeSdk (2.67M)                               | Microsoft.Net.UWPCoreRuntimeSdk (2.66M)
2    | runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk (2.69M)             | runtime.win10-x86.Microsoft.Net.UWPCoreRuntimeSdk (2.65M)
3    | runtime.win10-x64.Microsoft.Net.UWPCoreRuntimeSdk (2.68M)             | runtime.win10-arm.Microsoft.Net.UWPCoreRuntimeSdk (2.68M)
4    | runtime.win10-x86.Microsoft.Net.UWPCoreRuntimeSdk (2.67M)             | runtime.win10-x64.Microsoft.Net.UWPCoreRuntimeSdk (2.66M)
5    | Microsoft.WindowsAzure.Caching (1.72M)                                | Microsoft.Windows.SDK.BuildTools (485)
6    | Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel (71.14M) | runtime.win10-arm64.Microsoft.Net.UWPCoreRuntimeSdk (1.71k)
7    | Microsoft.NET.Test.Sdk (135.01M)                                      |
8    | System.Security.Principal.Windows (297.69M)                           |
9    | Microsoft.Win32.Registry (279.16M)                                    |
10   | Microsoft.IdentityModel.Logging (169.91M)                             |
</pre>

#### `microsoft toolkit high performance`

<pre>
Rank | <a href="https://www.nuget.org/packages?q=microsoft+toolkit+high+performance">Control (15.68k hits)</a>                                          | Treatment (3 hits)
---- | -------------------------------------------------------------- | --------------------------------------------------------------
1    | System.Diagnostics.PerformanceCounter (50.77M)                 | Microsoft.Toolkit.HighPerformance (1.25k)
2    | Microsoft.ApplicationInsights.PerfCounterCollector (71.78M)    | Uno.Microsoft.Toolkit.HighPerformance (234)
3    | Microsoft.CSharp (321.19M)                                     | NevronOpenVision (4.52k)
4    | Microsoft.NETCore.Platforms (494.33M)                          |
5    | Microsoft.Extensions.Options (487.92M)                         |
6    | Microsoft.Extensions.Primitives (546.8M)                       |
7    | Microsoft.Extensions.DependencyInjection.Abstractions (561.9M) |
8    | Microsoft.Extensions.Logging.Abstractions (492.5M)             |
9    | Microsoft.Extensions.Configuration.Abstractions (489.7M)       |
10   | Microsoft.Win32.Registry (279.16M)                             |
</pre>

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/SearchService/AzureSearchTextBuilder.cs [hunk 1]
```diff
@@ -9,11 +10,11 @@ namespace NuGet.Services.AzureSearch.SearchService
...
     public partial class SearchTextBuilder
     {
-        private enum TermPrefix
+        private enum Operator
         {
             None, // Default Lucene behavior, essentially "OR"
-            And, // "AND" / "&&" / "+"
-            Not, // "NOT" / "!" / "-"
+            Required, // "+" operator
+            Prohbit, // "-" operator
```
Comment (loic-sharma) [inline src/NuGet.Services.AzureSearch/SearchService/AzureSearchTextBuilder.cs:17]:
```suggestion
            Prohibit, // "-" operator
```

File: tests/NuGet.Services.AzureSearch.FunctionalTests/Support/TokenizationData.cs [hunk 2]
```diff
@@ -115,6 +115,13 @@ public static class TokenizationData
                 {
                     "once", "upon", "time", "little", "test", "case"
                 }
+            },
+            {
+                "ONCE UPON A TIME, THERE WAS A YELLING TEST-CASE!",
```
Comment (loic-sharma) [inline tests/NuGet.Services.AzureSearch.FunctionalTests/Support/TokenizationData.cs:120]: +1 for sassiness

File: src/NuGet.Services.AzureSearch/SearchService/AzureSearchTextBuilder.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/AzureSearchTextBuilder.cs]: Good catch.

File: src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs [hunk 3]
```diff
@@ -174,57 +177,78 @@ public SearchText Build(ParsedQuery parsed)
...
+                // that a term that the user adds to their search text is effective. In general, if tokens are optional,
+                // any score boost on, say, download count can cause highly popular but largely irrelevant packages to
+                // appear at the top. For the last token, allow a prefix match to support instant search scenarios.
+                var separatorTokens = unscopedTerms.SelectMany(TokenizeWithSeparators).ToList();
+                var uniqueSeparatorTokens = separatorTokens.ToHashSet();
+                foreach (var token in uniqueSeparatorTokens)
                 {
-                    builder.AppendBoostIfMatchAllTerms(unscopedTerms, _options.Value.MatchAllTermsBoost);
-                }
+                    var isLastToken = token == separatorTokens.Last();
```
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs:198]: I think in theory you are right, but I can't think of any patterns that would really need that. Also, we have the "namespace" booster to help with this case. I'll look into an elegant way to handle this case and add it if I find it.

**Review by loic-sharma — APPROVED** [review]
Code changes look good, excited to play around with this :)

File: src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (joelverhagen) [reply src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs:208]: Resolved.

**Comment by loic-sharma** [comment]
Joel deployed this to PROD PERF. If you want to play around with the new results in Visual Studio, here's my service index that points to PROD PERF search: https://loshardev0.blob.core.windows.net/api/prod-perf.json

**Comment by loic-sharma** [comment]
I played around with the search service, overall this feels better:

1. Queries with many terms appear to have improved. New search results has fewer "random" results that matched a single popular term.
1. Prefix matching seems to be improved. NUnit and NLogger are now results if you search for `n`. Longer prefix searches like `StackExchang` or `data protectio` also have better results

That said, the previous search service was more resilient to "bad" queries for popular packages. For example:

1. `newton soft` - Newtonsoft.Json is no longer in results
1. `azur storage queues` - No results, previously this query did well
1. `CloudStorageAccount` - No results, previously this query matched Azure Blob Storage packages

This seems like a good tradeoff to me 👍

**Comment by joelverhagen** [comment]
> 1. `newton soft` - Newtonsoft.Json is no longer in results

I guess this would need shingling on the query parsing side. This is unfortunate because as a human it seems extremely obvious that Newtonsoft.Json should be matched. I had an idea to make a concatenated version of simple `space separated tokens` an allowed alternative. It would help the `common mark` case as well.

> 1. `azur storage queues` - No results, previously this query did well

One of my attempts was to allow prefix on all tokens. I was able to get good results without such an extreme change so I moved away from it.

> 1. `CloudStorageAccount` - No results, previously this query matched Azure Blob Storage packages

This is a tough one. I think the best option would be to recommend the Azure Blob Storage package owners to add this tag, if this query is very popular.

Really great finds. We'll keep note of these here and investigate solutions if they prove to be important in the A/B test or long run.

**Pull request merged**